src/index/CompressedRelation.h

```diff
@@ -466,9 +476,33 @@ class CompressedRelationReader {
 
   // The specification of scan, together with the blocks on which this scan is
   // to be performed.
+  //
+  // Brief explanation of `ScanSpecAndBlocks` constructor.
+  // (1) The passed `ScanSpecification` remains as it is and is moved into
+  // member variable `scanSpec_`.
+  // (2) Member `blockMetadata_` is set to the `BlockMetadataRanges` computed
+  // via `getRelevantBlocks` for the provided `ScanSpecification` and
+  // `BlockMetadataRanges`.
+  // (3) `sizeBlockMetadata_` represents the number of
+  // `CompressedBlockMetadata` values contained over all subranges in member
+  // `BlockMetadataRanges_ blockMetadata_`.
   struct ScanSpecAndBlocks {
     ScanSpecification scanSpec_;
-    const std::span<const CompressedBlockMetadata> blockMetadata_;
+    BlockMetadataRanges blockMetadata_;
+    size_t sizeBlockMetadata_;
+
+    ScanSpecAndBlocks(ScanSpecification scanSpec,
+                      const BlockMetadataRanges& blockMetadataRanges)
+        : scanSpec_(std::move(scanSpec)),
+          blockMetadata_(getRelevantBlocks(scanSpec_, blockMetadataRanges)),
+          sizeBlockMetadata_(getNumberOfBlockMetadataValues(blockMetadata_)) {}
+
```

Comment:
You have the invariant, that the blocks that are passed in are all relevant for the given relation, and that they are sorted and so on and so on.
You should either establish or check all these invariants in the constructor
(same as in your prefiltering operations).